network/p2p/p2pnode/libp2pNode_test.go

```diff
@@ -242,3 +249,168 @@ func TestNode_HasSubscription(t *testing.T) {
+			case strings.Contains(message, "retrying peer dialing"):
+				dialPeerRetries.Inc()
+			}
+		}
+	})
+	logger := zerolog.New(os.Stdout).Level(zerolog.InfoLevel).Hook(hook)
+	idProvider := mockmodule.NewIdentityProvider(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	signalerCtx := irrecoverable.NewMockSignalerContext(t, ctx)
+
+	sporkID := unittest.IdentifierFixture()
+
+	sender, id1 := p2ptest.NodeFixture(
+		t,
+		sporkID,
+		t.Name(),
+		p2ptest.WithConnectionGater(testutils.NewConnectionGater(idProvider, func(pid peer.ID) error {
+			// avoid connection gating outbound messages on sender
+			return nil
+		})),
+		// add very small delay so that when the sender attempts to create multiple streams
+		// the func fails fast before the first routine can finish the peer dialing retries
+		// this prevents us from making another call to dial peer
+		p2ptest.WithCreateStreamRetryDelay(10*time.Millisecond),
+		p2ptest.WithLogger(logger))
+
+	receiver, id2 := p2ptest.NodeFixture(
+		t,
+		sporkID,
+		t.Name(),
+		p2ptest.WithConnectionGater(testutils.NewConnectionGater(idProvider, func(pid peer.ID) error {
+			// connection gate all incoming connections forcing the senders unicast manager to perform retries
+			return fmt.Errorf("gate keep")
+		})),
+		p2ptest.WithCreateStreamRetryDelay(10*time.Millisecond),
+		p2ptest.WithLogger(logger))
+
+	idProvider.On("ByPeerID", sender.Host().ID()).Return(&id1, true).Maybe()
+	idProvider.On("ByPeerID", receiver.Host().ID()).Return(&id2, true).Maybe()
+
+	p2ptest.StartNodes(t, signalerCtx, []p2p.LibP2PNode{sender, receiver}, 100*time.Millisecond)
+	defer p2ptest.StopNodes(t, []p2p.LibP2PNode{sender, receiver}, cancel, 100*time.Millisecond)
+
+	pInfo, err := utils.PeerAddressInfo(id1)
+	require.NoError(t, err)
+	sender.Host().Peerstore().AddAddrs(pInfo.ID, pInfo.Addrs, peerstore.AddressTTL)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	// attempt to create two concurrent streams
+	go func() {
+		defer wg.Done()
+		_, err := sender.CreateStream(ctx, receiver.Host().ID())
+		require.Error(t, err)
+	}()
+	go func() {
+		defer wg.Done()
+		_, err := sender.CreateStream(ctx, receiver.Host().ID())
+		require.Error(t, err)
+	}()
+	wg.Wait()
+
+	// we expect a single routine to start attempting to dial thus the number of retries
+	// before failure should be at most p2pnode.MaxConnectAttempt
+	expectedNumOfDialRetries := int64(p2pnode.MaxConnectAttempt)
+	// we expect the second routine to retry creating a stream p2pnode.MaxConnectAttempt when dialing is in progress
+	expectedCreateStreamRetries := int64(p2pnode.MaxConnectAttempt)
+	require.Equal(t, expectedNumOfDialRetries, dialPeerRetries.Load(), fmt.Sprintf("expected %d dial peer retries got %d", expectedNumOfDialRetries, dialPeerRetries.Load()))
+	require.Equal(t, expectedCreateStreamRetries, createStreamRetries.Load(), fmt.Sprintf("expected %d dial peer retries got %d", expectedCreateStreamRetries, createStreamRetries.Load()))
+}
```

Comment:
How does the test ensure that both create streams occur concurrently? Is there a possibility of flaky behavior if, for example, the first create stream call's dial attempt goes through before the second create stream call's goroutine kicks in, resulting in the second call not encountering any dial-in-progress error?

Reply:
We are forcing the first create stream call to kick off dialing retries because we are not adding the peer address info of node2 to node1 . So when node1 attempts to create a stream to node2 it will always encounter the "no addresses" error from libp2p forcing the dialing retries. Because of this the second call to create stream is guaranteed to always encounter the dial in progress error and attempt create stream retries.

Reply:
Thanks for the explanation, looks fine for now, but please create a follow-up issue to later address the detect immutable error states and fail fast upon facing one. For example, a `no addresses` error looks like an immutable error state, i.e., very unlikely that we won't face the same error upon the next retrial. In such situations, we better give up the stream/connection creation promptly. This yields both an efficient as well as a secure approach (see this for example https://github.com/onflow/flow-go/pull/3902#discussion_r1120971913).